### PR TITLE
Gutenberg 3.1

### PR DESCRIPTION
**Pull request by apardyl**
- Add basic print job control and logging capabilities,
- Handle jobs in async with Celery,
- Implement basic IPP 1.1, IPP 2.0 and IPP Everywhere support,
- Replace CAS authentication with Open ID Connect.

### CODE REVIEW

File: ipp/proto.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+class AttributeGroup(IppFieldsStruct):
+    _filter = None
+    _permanent_members = []
+    _default_filter = ['all']
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def write_to(self, writable, requested_attrs: List[str] = None):
```
[the remark this answers is not in the document]
Comment (apardyl) [reply ipp/proto.py]: Done.

File: ipp/proto.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
+import math
+import struct
+from abc import ABC
+from collections import OrderedDict
+from datetime import datetime
+from struct import Struct
+from typing import Tuple, List, Type, Any, Iterable
```
[the remark this answers is not in the document]
Comment (apardyl) [reply ipp/proto.py]: Done.

File: printing/printing.py [hunk 3]
```diff
@@ -80,9 +32,34 @@ def generate_hp500_options(copy_number: int, pages_to_print: str,
...
-        generate_options(**options))
-    cleanup()
+@shared_task
+def print_file(file_path, file_format, job_id):
+    job = PrintJob.objects.get(id=job_id)
+    job.status = JobStatus.PROCESSING
+    job.save()
+    tmpdir = None
+    try:
+        tmpdir = tempfile.mkdtemp()
```
[the remark this answers is not in the document]
Comment (apardyl) [reply printing/printing.py]: good point. done.

File: printing/printing.py [hunk 4]
```diff
@@ -80,9 +32,34 @@ def generate_hp500_options(copy_number: int, pages_to_print: str,
...
+        subprocess.check_call(
+            ['lp', out] + ['-d', settings.PRINTER_NAME] +
+            generate_options(job.properties))
+        shutil.rmtree(tmpdir)
+        job.status = JobStatus.COMPLETED
+        job.finished = timezone.now()
+        job.save()
+    except Exception as ex:
+        job.status = JobStatus.ERROR
+        job.status_reason = repr(ex)[:1999]
```
[the remark this answers is not in the document]
Comment (apardyl) [reply printing/printing.py]: Removed the max_length constraint on the field instead.

File: sandbox.sh [hunk 5]
```diff
@@ -6,12 +6,13 @@ bwrap --ro-bind / / \
       --tmpfs /home \
       --tmpfs /run \
       --tmpfs /tmp \
-      --tmpfs /root/.config/libreoffice \
```
[the remark this answers is not in the document]
Comment (apardyl) [reply sandbox.sh:9]: Oops. forgot to add it back. On systems, where /root is read protected sandbox failed when run by a non-root user as it couldn't mount anything on a destination it couldn't access. I replaced it with mounting tmpfs on /root and creating /root/.config/libreoffice dir (not sure if the last thing is needed, it seems to work without it).

File: ipp/service.py [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        return uri.replace('http', 'ipp')
+
+    def _get_job_uri(self, request: IppRequest, job_id: int):
+        uri = request.http_request.build_absolute_uri(
+            reverse('ipp_endpoint', kwargs={'token': self.user.api_key, 'rel_path': 'job/{}'.format(job_id)}))
+        return uri.replace('http', 'ipp')
+
+    def get_printer_attrs(self, request: IppRequest) -> IppResponse:
+        operation = request.read_group(GetPrinterAttributesRequestOperationGroup)
+        # logger.debug("GetPrinterAttrs:\n" + str(operation))
```
[the remark this answers is not in the document]
Comment (apardyl) [reply ipp/service.py:148]: Done.

File: sandbox.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (m4tx) [reply sandbox.sh:9]: Yes, I believe that LibreOffice will create `~/.config/libreoffice` directory gracefully when not present. Anyways, being extra safe by creating it by ourselves seems fine. LGTM.

**Review by m4tx — APPROVED** [review]

**Pull request merged**